Error opening/creating journal files over remote connection

If a remote development method is used (ssh in this case), new journal file creation gives an error, but apparently succeeds as subsequent access to file location works fine.

Repro:

- Open remote development connection to any Linux server over ssh with vscode-journal extension installed
- Ctrl+Shift+J -> Today

If file does not already exist, following error is logged:

    [09:58:07.054] [ERROR] 
    [09:58:07.059] [ERROR] Failed to process input. See Exception below.
    Error: cannot open vscode-remote://ssh-remote%2Bhost/path/to/Journal/2019/10/24.md. Detail: Unable to read file (Error: File not found (vscode-remote://ssh-remote+host/path/to/Journal/2019/10/24.md))
        at n.then.e (file:///usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:4707:361)
        at process._tickCallback (internal/process/next_tick.js:68:7)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,9 @@
 
 ## Unreleased
 
+### Fixed
+* [Issue #51](https://github.com/pajoma/vscode-journal/issues/51) Remote SSH and WSL Remote no longer surface a "File not found" error toast on first-time creation of an entry, weekly page, or note. Replaced the open-before-create antipattern in `Reader.loadEntryForDay`, `Reader.loadEntryForWeek`, and `LoadNotes.loadNote` with a stat-first check (new `fileExists` helper in `src/util/fs-exists.ts`) and migrated the three methods to native `async/await`.
+
 ### Changed
 * Bumped VS Code engine baseline to `^1.118.0` (was `^1.94.0`). `@types/vscode` and toolchain (TypeScript 5.9, ESLint 9.39, esbuild 0.28, `@vscode/test-cli` 0.0.12, `@vscode/test-electron` 2.5, Mocha 11, `@typescript-eslint/*` 8.59) bumped to current stable.
 * Migrated translations from the custom `src/ext/messages.json` system to the stable `vscode.l10n` API. Manifest strings live in `package.nls*.json`; runtime strings in `l10n/bundle.l10n*.json`. All 11 locales preserved (en, de, fr, es, it, pt, nl, ru, zh, ja, ar). Fixed a moment.js escape bug in the Spanish locale carried over from `messages.json`.

--- a/PLAN.md
+++ b/PLAN.md
@@ -57,6 +57,7 @@ Make sure that, when running on a remote host, the extension can still access th
 - [x] `os.homedir()` in `conf.ts`: No change needed — with `extensionKind: ["workspace"]` the extension runs on the remote host, so `os.homedir()` correctly returns the remote home directory where the journal resides
 - [x] Replace `untitled:` URI scheme in `writer.ts:createSaveLoadTextDocument()` with `vscode.workspace.fs.writeFile()` + `vscode.workspace.openTextDocument()`
 - [x] Add integration test verifying file creation via `vscode.workspace.fs`
+- [x] **Followup #51**: First-time entry/week/note creation on Remote SSH and WSL still surfaced VS Code's error toast because all three call sites (`Reader.loadEntryForDay`, `Reader.loadEntryForWeek`, `LoadNotes.loadNote`) used the open-before-create antipattern and relied on prefix-matching the error message text. Replaced with stat-first via a dedicated `fileExists(uri)` helper in `src/util/fs-exists.ts`. Migrated the three methods from `new Promise(...)` wrappers to native `async/await` while there. Regression tests in `src/test/suite/issue-51-remote-create.test.ts`. See [docs/specs/2026-05-14-fix-51-remote-file-creation.md](docs/specs/2026-05-14-fix-51-remote-file-creation.md) and [docs/plans/2026-05-14-fix-51-remote-file-creation.md](docs/plans/2026-05-14-fix-51-remote-file-creation.md).
 
 ---
 

--- a/docs/plans/2026-05-14-fix-51-remote-file-creation.md
+++ b/docs/plans/2026-05-14-fix-51-remote-file-creation.md
@@ -1,0 +1,289 @@
+# Issue #51 — Implementation Plan: Remote file creation error
+
+> **Spec:** [docs/specs/2026-05-14-fix-51-remote-file-creation.md](../specs/2026-05-14-fix-51-remote-file-creation.md)
+> **Issue:** [pajoma/vscode-journal#51](https://github.com/pajoma/vscode-journal/issues/51)
+> **Branch:** `51-error-openingcreating-journal-files-over-remote-connection`
+> **Created:** 2026-05-14
+
+## Approach
+
+Replace the speculative "call `openTextDocument`, catch failure, create" pattern with a deterministic "stat first, then open OR create" pattern in three methods:
+
+1. `Reader.loadEntryForDay`
+2. `Reader.loadEntryForWeek`
+3. `LoadNotes.loadNote`
+
+A small dedicated helper `fileExists(uri)` in its own module (`src/util/fs-exists.ts`) wraps `vscode.workspace.fs.stat` and converts `FileSystemError.FileNotFound` to `false`. Anything else propagates so VS Code surfaces an error dialog (permission denied, transport unavailable, etc.). No reliance on prefix-matching error messages.
+
+While each method is rewritten, the existing `new Promise((resolve, reject) => …)` wrapper around the body is removed and the function body becomes a flat `async`/`await` sequence. This was previously out of scope but pulled in by the approval comment because the wrappers make the new linear flow noticeably harder to read.
+
+Trade-offs:
+- The diff touches three methods, one new helper file, and adds regression tests. The Promise flattening grows the diff size but keeps each method idiomatic — the alternative (leave the wrappers, just patch the body) leaves a worse code shape behind.
+- The change cannot be unit-tested with mocks alone — the assertion ("no error logged on first create") needs a real `Logger` and real `vscode.workspace.fs`. Tests run in the Extension Host (already the project's convention via `@vscode/test-cli`). Remote-host verification stays manual.
+- Detection technique: TDD. Write each regression test first, watch it fail on current code, refactor, watch it pass.
+
+## Steps
+
+### Step 1 — Branch + baseline
+
+- Branch `51-error-openingcreating-journal-files-over-remote-connection` already checked out. Spec at commit `7d61846`.
+- Run `npm install` (toolchain matches `develop` HEAD `0744437` after PR #176 plus the #170 commits merged via PR #178 on this branch — same deps as the #170 work).
+- Run `npm run check`. Confirm baseline is clean before introducing the fix.
+
+### Step 2 — Failing regression tests (TDD)
+
+New file: `src/test/suite/issue-51-remote-create.test.ts`. Pattern matches `phase1-regression.test.ts` — real `vscode.workspace.fs`, `TestLogger` in capturing mode so the assertion "no error logged" is concrete.
+
+Test scenarios (named) — see "Test scenarios" section below for the full plain-language list. Each test:
+- Creates a unique tmp directory for the test (under `os.tmpdir()`), avoiding collision with the configured journal base.
+- Constructs a real `Ctrl`, points the relevant config at the tmp path, invokes the method.
+- Asserts: returned `TextDocument.uri.fsPath` matches; `getText()` contains the template header; `TestLogger.errors` is empty.
+- Tears down by deleting the tmp directory via `vscode.workspace.fs.delete`.
+
+Helper test file extension: extend `TestLogger` (in `src/test/test-logger.ts`) with an `errors: string[]` accumulator if it does not already capture `printError`/`error` calls. Inspect first; add only if missing.
+
+Run `npm test -- --grep "#51"`. Confirm tests fail (expected: error is logged on first-time creation).
+
+### Step 3 — Add `fileExists` helper module
+
+New file: `src/util/fs-exists.ts`.
+
+```typescript
+import * as vscode from 'vscode';
+
+export async function fileExists(uri: vscode.Uri): Promise<boolean> {
+    try {
+        await vscode.workspace.fs.stat(uri);
+        return true;
+    } catch (err) {
+        if (err instanceof vscode.FileSystemError && err.code === 'FileNotFound') {
+            return false;
+        }
+        throw err;
+    }
+}
+```
+
+Tests for the helper go in the same `issue-51-remote-create.test.ts`:
+- Missing path → returns `false`, no log.
+- Existing path → returns `true`.
+- Re-throws non-FileNotFound (simulated by passing an invalid/inaccessible URI — exact technique TBD during impl; if a clean simulation isn't available in the test env, document this as a manual check and rely on the contract in the helper code itself).
+
+Register the helper in `src/util/index.ts` so it is reachable via `J.Util.fileExists` (the project's namespace barrel pattern — even though new code should prefer named imports per CLAUDE.md, the existing callers in `Reader` import via `J.Util.*`, so consistency wins for this PR).
+
+### Step 4 — Refactor `Reader.loadEntryForDay`
+
+File: `src/actions/reader.ts` (lines 97–143).
+
+Replace the entire `Promise<vscode.TextDocument>` body with:
+
+```typescript
+public async loadEntryForDay(date: Date): Promise<vscode.TextDocument> {
+    if (J.Util.isNullOrUndefined(date) || date!.toString().includes("Invalid")) {
+        throw new Error("Invalid date");
+    }
+    this.ctrl.logger.trace("Entering loadEntryforDate() in actions/reader.ts for date " + date.toISOString());
+
+    const [pathname, filename] = await Promise.all([
+        this.ctrl.config.getResolvedEntryPath(date),
+        this.ctrl.config.getEntryFilePattern(date),
+    ]);
+    const path = J.Util.resolvePath(pathname.value!, filename.value!);
+    const uri = vscode.Uri.file(path);
+
+    const exists = await J.Util.fileExists(uri);
+    const doc = exists
+        ? await this.ctrl.ui.openDocument(path)
+        : await this.ctrl.writer.createEntryForPath(path, date);
+
+    this.ctrl.logger.debug("loadEntryForDate() - Loaded file in:", doc.uri.toString());
+
+    // fire-and-forget link injection (preserves existing side-effect)
+    new J.Provider.SyncNoteLinks(this.ctrl).injectAttachementLinks(doc, date)
+        .finally(() => this.ctrl.logger.trace("Scanning notes completed"));
+
+    return doc;
+}
+```
+
+Behavioral notes:
+- Removed prefix-matching catch block (`reader.ts:118–124`). With stat-first, no spurious failure can reach a catch on the happy path.
+- `J.Util.fileExists` propagates non-FileNotFound errors → bubbles out of `loadEntryForDay` → caller (the smart-input command) handles. No try/catch here; the caller already wraps with its own error handling that surfaces the dialog.
+- "Invalid date" branch now throws instead of resolving via reject — same observable behavior since the original caller treats both as a rejection.
+
+### Step 5 — Refactor `Reader.loadEntryForWeek`
+
+File: `src/actions/reader.ts` (lines 54–86).
+
+Mirror image of step 4 (no link injection — weekly entries do not call `SyncNoteLinks`). Body becomes:
+
+```typescript
+public async loadEntryForWeek(week: Number): Promise<vscode.TextDocument> {
+    this.ctrl.logger.trace("Entering loadEntryForWeek() in actions/reader.ts for week " + week);
+
+    const [pathname, filename] = await Promise.all([
+        this.ctrl.config.getWeekPathPattern(week),
+        this.ctrl.config.getWeekFilePattern(week),
+    ]);
+    const path = J.Util.resolvePath(pathname.value!, filename.value!);
+    const uri = vscode.Uri.file(path);
+
+    const exists = await J.Util.fileExists(uri);
+    const doc = exists
+        ? await this.ctrl.ui.openDocument(path)
+        : await this.ctrl.writer.createWeeklyForPath(path, week);
+
+    this.ctrl.logger.debug("loadEntryForWeek() - Loaded file in:", doc.uri.toString());
+    return doc;
+}
+```
+
+This is the call site that previously *only* checked `cannot open file:` (not `cannot open vscode-remote:`), so it carried the loudest version of the bug on remote.
+
+### Step 6 — Refactor `LoadNotes.loadNote`
+
+File: `src/provider/features/load-note.ts` (lines 42–60).
+
+```typescript
+public async loadNote(path: string, content: string): Promise<vscode.TextDocument> {
+    this.ctrl.logger.trace("Entering loadNote() in features/load-note.ts for path: ", path);
+
+    const uri = vscode.Uri.file(path);
+    const exists = await J.Util.fileExists(uri);
+    if (exists) {
+        return this.ctrl.ui.openDocument(path);
+    }
+    return this.ctrl.writer.createSaveLoadTextDocument(path, content);
+}
+```
+
+`loadWithPath` (lines 21–32) is unchanged; it already calls `loadNote` and `loadEntryForInput` with `async/await` at the outer level.
+
+### Step 7 — Run tests, iterate
+
+- `npm test -- --grep "#51"` should now pass.
+- `npm run check` — lint + compile + full suite must be green.
+- If any unrelated test breaks (e.g. commands-entry stubs `loadEntryForInput` at the controller seam — unaffected), investigate; do not paper over.
+
+### Step 8 — Manual verification
+
+- Local: `F5` (Run Extension), trigger `Ctrl+Shift+J → Today` on a date with no existing file, confirm clean open, no error toast. Repeat for weekly entry and a new note.
+- Remote SSH: open a workspace on a Linux host via Remote-SSH, install the dev build (sideload via `vsce package` and `code --install-extension`), repeat the three actions. Capture the OutputChannel content for the PR description as evidence.
+- WSL Remote: same as SSH, on a Windows host with WSL. Same evidence.
+- Pre-existing-file path: trigger `Today` after the entry is already created — confirm it opens the same document, does not overwrite.
+
+If remote sideload is impractical (e.g. no spare remote host), substitute by manually testing against an SMB/9P-mounted directory that mimics async FS latency. Document this substitution in the PR description if used.
+
+### Step 9 — PR + changelog
+
+- Branch already exists. Commit history: spec (`d5c140d`) → spec amendments (`7d61846`) → plan (this doc) → tests → helper → reader refactor → notes refactor → manual evidence.
+- Open PR against `develop`. Title: `fix: stat-first file creation on remote workspaces (#51)`.
+- PR body links: issue, spec doc, plan doc, the OutputChannel evidence from step 8.
+- Update `CHANGELOG.md` under the unreleased section with one line referencing #51.
+- Update `PLAN.md` Phase 1.3 to record that the user-visible payoff landed.
+- Post on the issue: "PR #Y opens — implementing approved plan." Nothing else.
+
+## Test scenarios
+
+### S1 — `Reader.loadEntryForDay` creates and opens, no log
+
+**Type:** integration (Extension Host, real FS, real Logger).
+**Acceptance link:** spec acceptance #1, #2 (no `[ERROR]` line).
+**Setup:** tmp dir; configure journal base to it; pick a date whose computed path does not exist.
+**Action:** `await ctrl.reader.loadEntryForDay(date)`.
+**Assert:** returned doc exists at expected path; content includes the template header; `TestLogger.errors.length === 0`.
+
+### S2 — `Reader.loadEntryForDay` opens existing file without overwrite
+
+**Type:** integration.
+**Setup:** tmp dir; pre-write a file with sentinel content at the computed path.
+**Action:** `await ctrl.reader.loadEntryForDay(date)`.
+**Assert:** sentinel content preserved; no log.
+
+### S3 — `Reader.loadEntryForWeek` creates and opens, no log
+
+**Type:** integration. Mirror of S1 for the week path.
+
+### S4 — `Reader.loadEntryForWeek` opens existing file without overwrite
+
+**Type:** integration. Mirror of S2.
+
+### S5 — `LoadNotes.loadNote` creates and opens, no log
+
+**Type:** integration. Mirror of S1 for the notes path.
+
+### S6 — `LoadNotes.loadNote` opens existing note without overwrite
+
+**Type:** integration. Mirror of S2 for the notes path.
+
+### S7 — `fileExists` helper: missing path
+
+**Type:** unit (with real `vscode.workspace.fs` — no genuine mocking needed; the helper is one tiny function and tests run in the Host anyway).
+**Action:** `await fileExists(vscode.Uri.file('/tmp/does-not-exist-' + Date.now()))`.
+**Assert:** returns `false`; no log.
+
+### S8 — `fileExists` helper: existing path
+
+**Type:** unit.
+**Setup:** tmp file written via `writeFile`.
+**Assert:** returns `true`.
+
+### S9 — Manual: Remote SSH first-time entry create
+
+**Type:** e2e (manual).
+**Acceptance link:** spec acceptance #2.
+**Setup:** Linux host via Remote-SSH; workspace where today's entry path is missing.
+**Action:** `Ctrl+Shift+J → Today`.
+**Assert:** entry opens immediately on first attempt; no notification toast; Journal OutputChannel has no `[ERROR]` line for the action.
+
+### S10 — Manual: Remote SSH weekly entry create
+
+**Type:** e2e (manual). Same shape as S9 with `Ctrl+Shift+J → Week`.
+
+### S11 — Manual: Remote SSH new note create
+
+**Type:** e2e (manual). Same shape as S9 with `note <title>` smart-input.
+
+### S12 — Manual: WSL Remote first-time entry create
+
+**Type:** e2e (manual). Same as S9 on WSL transport.
+
+### S13 — Manual: WSL Remote weekly + notes
+
+**Type:** e2e (manual). Same as S10 + S11 on WSL.
+
+### S14 — Manual: Local entry create regression
+
+**Type:** e2e (manual).
+**Acceptance link:** spec acceptance #4 (local unchanged).
+**Setup:** Local workspace; entry path missing.
+**Action:** `Ctrl+Shift+J → Today`.
+**Assert:** same observable behavior as before this PR (clean open).
+
+## Dependencies
+
+None. No cross-issue blockers, no cross-repo branches to land first.
+
+Soft dependency: the branch already carries the merged #170 commits (PR #178 was opened from `170-…` and merged into this branch's ancestry). `npm install` may show no changes; if so, proceed.
+
+## Risk
+
+- **R1: `vscode.FileSystemError.code` value coupling.** The helper checks `err.code === 'FileNotFound'`. If a future VS Code version renames or returns the error differently, the helper will misroute to "re-throw" and the create path will never trigger. *Mitigation:* S7 covers the missing-file case; if the API changes, the test breaks immediately, before users do. Also, the VS Code API surface guarantees `FileSystemError.FileNotFound` is the canonical "not found" code per the docs.
+
+- **R2: Race against parallel triggers.** If two `Today` invocations happen simultaneously (unlikely — single user, single shortcut), stat-then-create has a TOCTOU window. *Mitigation:* `Writer.createSaveLoadTextDocument` uses `vscode.workspace.fs.writeFile` which is idempotent overwrite — the second call would overwrite the first's empty header. The risk is purely theoretical for this UX; not worth a lock.
+
+- **R3: Removing the `printError` log call hides legitimate errors.** In the old reader code, non-FileNotFound errors during `openDocument` went through `printError`. With stat-first, non-FileNotFound errors propagate as exceptions to the caller. *Mitigation:* the caller (`provider/commands/show-entry-for-today.ts` and similar) already logs and surfaces — verify during impl. If a caller swallows silently, fix that caller too (in scope as part of this PR if discovered).
+
+- **R4: Test environment cannot simulate non-FileNotFound errors.** The "re-throws on other errors" branch may end up untested. *Mitigation:* document the contract in the helper code; add a TODO test if a clean simulation technique appears later. The risk is bounded because the helper is one-screen tiny.
+
+- **R5: `J.Util.fileExists` barrel export creates a new entry point that diverges from the spec's wording.** Spec says "own module". The barrel re-export is a project convention, not a contradiction — the file is still its own module. *Mitigation:* mentioned in step 3; consistent with `J.Util.resolvePath` etc.
+
+## Rollback
+
+The change is contained: revert the merge commit (or the squashed PR commit) on `develop`. No data migration, no config schema change, no marketplace-side rollback needed. Pre-existing journal files are unaffected (the change only alters how new files are created).
+
+If a partial issue surfaces post-merge (e.g. one of the three flows still logs on first create), the fix can be a follow-up PR rather than a full revert — the three methods are independent.
+
+## Reference spec
+
+`docs/specs/2026-05-14-fix-51-remote-file-creation.md` (committed at `d5c140d`, amended at `7d61846`).

--- a/docs/specs/2026-05-14-fix-51-remote-file-creation.md
+++ b/docs/specs/2026-05-14-fix-51-remote-file-creation.md
@@ -1,0 +1,111 @@
+# Issue #51 — Error opening/creating journal files over remote connection
+
+> **Issue:** [pajoma/vscode-journal#51](https://github.com/pajoma/vscode-journal/issues/51)
+> **Branch:** `51-error-openingcreating-journal-files-over-remote-connection`
+> **Created:** 2026-05-14
+
+## Goal
+
+When the user triggers a journal action (`journal.today`, `journal.day`, weekly entry, or note creation) in a Remote SSH or WSL workspace and the target file does not yet exist, the extension must create the file and open it without surfacing any error to the user. No `[ERROR]` line in the output channel, no VS Code error notification toast, no requirement to retry the command.
+
+## Why now
+
+Open since 2019 (issue #51) with thumbs-up from multiple users on the `1.1.0` milestone. The user experience is "press shortcut → see scary error → press shortcut again → it works." Even when no work is lost, the error breaks trust and is the first thing remote users hit on day one. Phase 1.3 of `PLAN.md` migrated raw `fs` calls to `vscode.workspace.fs` and the manifest now declares `extensionKind: ["workspace"]`, so the infrastructure is in place to make remote work cleanly — but the open-before-create code path was not touched and still produces the symptom.
+
+## Reproduction status
+
+**Unknown on the current `develop` HEAD.** The fix MUST start with a verified reproduction (or a definitive "the symptom is gone for case X but still present for case Y") in the implementation plan. This spec lists the code paths that still carry the broken pattern — if any of them are exercised, the bug is still live.
+
+## Root cause (analysis)
+
+All three "open or create" code paths use the same shape: call `vscode.workspace.openTextDocument(uri)` first, catch the failure, then `createSaveLoadTextDocument`. On a remote workspace, `openTextDocument` on a missing file surfaces a VS Code internal error and rejects with a message string that starts with `cannot open vscode-remote:` (or `cannot open file:` locally).
+
+The three call sites:
+
+1. **`src/actions/reader.ts:97` — `Reader.loadEntryForDay`.** Catch at line 118–124 filters on both prefixes (`cannot open file:` AND `cannot open vscode-remote:`), so the *logger* stays quiet on the happy path, but the VS Code error toast from the failed `openTextDocument` is independent of our catch and still reaches the user.
+2. **`src/actions/reader.ts:54` — `Reader.loadEntryForWeek`.** Catch at line 68–74 only filters `cannot open file:`. On a remote workspace, the week entry path logs `[ERROR]` AND surfaces the toast.
+3. **`src/provider/features/load-note.ts:42` — `LoadNotes.loadNote`.** Catch at line 50 does not filter anything; it falls through to `createSaveLoadTextDocument` but still logs `error` and the toast appears.
+
+Common downstream: `Writer.createSaveLoadTextDocument` (`src/actions/writer.ts:117`) is the actual create+open. It uses `vscode.workspace.fs.writeFile(...)` then `vscode.workspace.openTextDocument(...)`, both of which are remote-aware. The create itself is not broken — only the speculative *open before create* is.
+
+The fix must replace "open and catch failure" with "check existence first, then open OR create." `vscode.workspace.fs.stat(uri)` is the silent existence check (rejects with a typed `FileSystemError` that does not bubble a notification).
+
+## Scope
+
+### In scope
+
+1. **Replace open-before-create with stat-first** in all three call sites:
+   - `Reader.loadEntryForDay` (entry path)
+   - `Reader.loadEntryForWeek` (weekly path)
+   - `LoadNotes.loadNote` (notes path)
+
+   Pattern: `await vscode.workspace.fs.stat(uri)` → if rejects with `FileNotFound`, call writer; otherwise call `openDocument`. No reliance on error-message prefix matching.
+
+2. **Remove the prefix-matching catch logic** at `reader.ts:118–124` and `reader.ts:68–74` once stat-first is in place. The fragile string check disappears with the pattern.
+
+3. **Confirm `Writer.createSaveLoadTextDocument` is remote-safe** — `vscode.Uri.file(path)` from a workspace-host extension resolves to the workspace filesystem, which on remote IS the remote FS. No change expected here, but verify with a regression test.
+
+4. **Regression coverage.** New unit/integration tests in `src/test/suite/` that:
+   - Drive `Reader.loadEntryForDay` against a non-existent path → asserts file is created, opened, and **no error is logged** via `Logger`.
+   - Drive `Reader.loadEntryForWeek` against a non-existent path → same assertion.
+   - Drive `LoadNotes.loadNote` against a non-existent path → same assertion.
+   - Pre-existing file → opens it, does NOT overwrite.
+
+   Tests run against the local workspace FS (the extension host's `vscode.workspace.fs` already abstracts local vs remote — the same code path that fails on remote also fails on local if the assertion is "no error logged"). Manual remote verification is still required for the acceptance check.
+
+### Out of scope
+
+- Refactoring `Reader` to drop the `new Promise((resolve, reject) => …)` wrappers around already-async code (`PLAN.md` Phase 2 owns the native `async/await` cleanup). The stat-first change will land *inside* the existing promise wrappers and a follow-up task can flatten them.
+- Replacing `vscode.Uri.file(...)` with `vscode.workspace.workspaceFolders[0].uri.with(...)` or similar workspace-relative URI construction. The current `Uri.file` form already works on remote because the extension runs as a workspace extension; verify, do not change.
+- Codespaces / dev containers (user scoped out — `Remote SSH` + `WSL Remote` only). Codespaces uses the same `vscode-remote:` family so the fix should incidentally cover it, but it is not part of the acceptance criteria.
+- The `LoadNotes.load → loadEntryForInput` recursive call (`load-note.ts:28`). It calls back into `Reader.loadEntryForDay`, which is already covered by item 1 — but the recursion means the fix for #1 must land before notes work fully on remote.
+
+## Acceptance criteria
+
+1. `npm run check` passes — lint clean, esbuild compiles, all existing tests pass, every new regression test from "In scope #4" passes on first run.
+2. Manual repro on **Remote SSH** (a Linux host over SSH, VS Code remote-ssh extension):
+   - Open a fresh workspace where the journal path resolves to a directory with no entry for today.
+   - Trigger `Ctrl+Shift+J` → `Today`.
+   - The new entry opens immediately on the first attempt. No error notification toast. The OutputChannel (`Journal` log) contains no `[ERROR]` line for the action.
+   - Repeat with `Ctrl+Shift+J` → `Week` (weekly entry creation).
+   - Repeat with `note new note title` (notes creation).
+3. Manual repro on **WSL Remote**: same three actions, same assertions.
+4. Existing local behavior is unchanged — opening today's entry on the local FS still works whether the file exists or not.
+5. No reintroduction of raw `fs` / `fs.promises` (PLAN.md Phase 1.3 invariant).
+
+## Entities / contracts touched
+
+- `src/actions/reader.ts`
+  - `loadEntryForDay(date: Date): Promise<vscode.TextDocument>` — rewrite the `Promise.all → .then → .catch` chain to a stat-first flow.
+  - `loadEntryForWeek(week: Number): Promise<vscode.TextDocument>` — same rewrite.
+- `src/provider/features/load-note.ts`
+  - `loadNote(path: string, content: string): Promise<vscode.TextDocument>` — replace open-and-catch with stat-first.
+- `src/util/paths.ts` (optional)
+  - There may be value in extracting a `existsViaFs(uri: vscode.Uri): Promise<boolean>` helper that wraps `vscode.workspace.fs.stat` with the `FileNotFound` swallow. `paths.ts:172` already uses `fs.stat` for a similar check; consolidate if it cleans up the call sites.
+- `src/actions/writer.ts`
+  - Likely unchanged. Inspect during implementation to confirm `createSaveLoadTextDocument` does not need a similar stat-first guard against accidental overwrite (current behavior: it WILL overwrite if the file exists, which is the intended semantics for the "create" path — callers are responsible for not calling it on an existing file). The stat-first refactor in callers preserves this invariant.
+- `src/test/suite/` — new test file (e.g. `reader-remote-create.test.ts`) or extensions to `phase1-regression.test.ts`.
+
+## Constraints
+
+- Must use `vscode.workspace.fs` for all FS operations (no `fs` / `fs.promises`).
+- Cannot rely on the error message text of `openTextDocument` — that is the current fragility.
+- Must not change the `Promise<vscode.TextDocument>` signatures of the three rewritten methods (callers in commands expect them).
+- Must preserve the side effect at `reader.ts:129` (`SyncNoteLinks.injectAttachementLinks`) — the link injection happens after open OR create, on every successful load.
+
+## Open questions
+
+1. Should `existsViaFs` live in `src/util/paths.ts` (already exports `resolvePath` and uses `fs.stat`) or in a new module? Decision punted to the plan step — the implementation will pick whichever keeps the diff smaller.
+2. The current catch in `loadEntryForDay` (line 118) discriminates between "open failed because file missing" (fall through to create) and "open failed for some other reason" (reject, log). With stat-first, a `stat` rejection that is NOT `FileNotFound` (e.g. permission denied) must still reject loudly. The implementation must preserve that branch.
+
+## Related issues
+
+- `PLAN.md` Phase 1.3 — `fs` → `vscode.workspace.fs` migration. This issue is the user-visible payoff of that phase.
+- No cross-repo dependencies. No `blocked-by:` / `blocks:` labels needed.
+
+## Reference
+
+- Issue body: `Ctrl+Shift+J → Today` on remote SSH produces `Error: cannot open vscode-remote://…/2019/10/24.md` followed by a successful next attempt.
+- Owner repro (in issue): same error in WSL.
+- Three additional `+1` comments from other users on the same milestone.
+- Current code: `src/actions/reader.ts:97`, `src/actions/reader.ts:54`, `src/provider/features/load-note.ts:42`, `src/actions/writer.ts:117`.

--- a/docs/specs/2026-05-14-fix-51-remote-file-creation.md
+++ b/docs/specs/2026-05-14-fix-51-remote-file-creation.md
@@ -3,6 +3,7 @@
 > **Issue:** [pajoma/vscode-journal#51](https://github.com/pajoma/vscode-journal/issues/51)
 > **Branch:** `51-error-openingcreating-journal-files-over-remote-connection`
 > **Created:** 2026-05-14
+> **Approved:** 2026-05-14 ([#issuecomment-4454118587](https://github.com/pajoma/vscode-journal/issues/51#issuecomment-4454118587)) — with three amendments folded into "In scope" #4–#6 below.
 
 ## Goal
 
@@ -45,20 +46,27 @@ The fix must replace "open and catch failure" with "check existence first, then 
 
 3. **Confirm `Writer.createSaveLoadTextDocument` is remote-safe** — `vscode.Uri.file(path)` from a workspace-host extension resolves to the workspace filesystem, which on remote IS the remote FS. No change expected here, but verify with a regression test.
 
-4. **Regression coverage.** New unit/integration tests in `src/test/suite/` that:
+4. **Flatten the `new Promise((resolve, reject) => …)` wrappers around already-async code** in the three methods being touched. Migrate to native `async/await` as the stat-first rewrite lands — both belong in the same diff because the new flow is naturally linear and the wrappers obscure it. Scope limited to `Reader.loadEntryForDay`, `Reader.loadEntryForWeek`, `LoadNotes.loadNote` (no opportunistic rewrites elsewhere). Aligns with PLAN.md Phase 2 direction.
+
+5. **Helper module for "does this file exist".** Add a small dedicated module (e.g. `src/util/fs-exists.ts`) exposing `fileExists(uri: vscode.Uri): Promise<boolean>` that wraps `vscode.workspace.fs.stat` and converts the `FileSystemError.FileNotFound` case into `false`. One file per module — do NOT inline the helper at the call sites. Non-`FileNotFound` errors propagate so the caller can surface a loud error dialog (see #6).
+
+6. **Loud rejection for non-`FileNotFound` errors.** A `stat` rejection with code `FileSystemError` other than `FileNotFound` (e.g. `NoPermissions`, `Unavailable`) must propagate up so the user sees an error dialog — not silently fall through to "create". Implementation must catch only `FileNotFound` and re-throw the rest. Logger records the original error.
+
+7. **Regression coverage.** New unit/integration tests in `src/test/suite/` that:
    - Drive `Reader.loadEntryForDay` against a non-existent path → asserts file is created, opened, and **no error is logged** via `Logger`.
    - Drive `Reader.loadEntryForWeek` against a non-existent path → same assertion.
    - Drive `LoadNotes.loadNote` against a non-existent path → same assertion.
    - Pre-existing file → opens it, does NOT overwrite.
+   - `fileExists` helper: returns `false` on missing file (no log), returns `true` on existing file, re-throws on non-FileNotFound errors.
 
    Tests run against the local workspace FS (the extension host's `vscode.workspace.fs` already abstracts local vs remote — the same code path that fails on remote also fails on local if the assertion is "no error logged"). Manual remote verification is still required for the acceptance check.
 
 ### Out of scope
 
-- Refactoring `Reader` to drop the `new Promise((resolve, reject) => …)` wrappers around already-async code (`PLAN.md` Phase 2 owns the native `async/await` cleanup). The stat-first change will land *inside* the existing promise wrappers and a follow-up task can flatten them.
 - Replacing `vscode.Uri.file(...)` with `vscode.workspace.workspaceFolders[0].uri.with(...)` or similar workspace-relative URI construction. The current `Uri.file` form already works on remote because the extension runs as a workspace extension; verify, do not change.
 - Codespaces / dev containers (user scoped out — `Remote SSH` + `WSL Remote` only). Codespaces uses the same `vscode-remote:` family so the fix should incidentally cover it, but it is not part of the acceptance criteria.
 - The `LoadNotes.load → loadEntryForInput` recursive call (`load-note.ts:28`). It calls back into `Reader.loadEntryForDay`, which is already covered by item 1 — but the recursion means the fix for #1 must land before notes work fully on remote.
+- Opportunistic `new Promise` flattening in *other* methods of `Reader`, `Writer`, `Inject` etc. (`saveDocument`, `createEntryForPath`, `createWeeklyForPath`, `loadEntryForInput`). Tempting to clean up the whole file, but stays out of this PR to keep the review surface small. PLAN.md Phase 2 still owns those.
 
 ## Acceptance criteria
 
@@ -95,8 +103,10 @@ The fix must replace "open and catch failure" with "check existence first, then 
 
 ## Open questions
 
-1. Should `existsViaFs` live in `src/util/paths.ts` (already exports `resolvePath` and uses `fs.stat`) or in a new module? Decision punted to the plan step — the implementation will pick whichever keeps the diff smaller.
-2. The current catch in `loadEntryForDay` (line 118) discriminates between "open failed because file missing" (fall through to create) and "open failed for some other reason" (reject, log). With stat-first, a `stat` rejection that is NOT `FileNotFound` (e.g. permission denied) must still reject loudly. The implementation must preserve that branch.
+Resolved in approval comment ([#issuecomment-4454118587](https://github.com/pajoma/vscode-journal/issues/51#issuecomment-4454118587)):
+1. ~~Helper location~~ → Own module, `src/util/fs-exists.ts` (one file per module, no inline).
+2. ~~Non-`FileNotFound` behavior~~ → Error dialog required. Helper re-throws non-FileNotFound; caller propagates so VS Code surfaces a notification.
+3. ~~Promise → async/await~~ → In scope for the three touched methods.
 
 ## Related issues
 

--- a/src/actions/reader.ts
+++ b/src/actions/reader.ts
@@ -52,96 +52,64 @@ export class Reader {
      * @param week the week of the current year
      */
     public async loadEntryForWeek(week: Number): Promise<vscode.TextDocument> {
-        return new Promise<vscode.TextDocument>((resolve, reject) => {
-            this.ctrl.logger.trace("Entering loadEntryForWeek() in actions/reader.ts for week " + week);
+        this.ctrl.logger.trace("Entering loadEntryForWeek() in actions/reader.ts for week " + week);
 
-            let path: string = "";
+        const [pathname, filename] = await Promise.all([
+            this.ctrl.config.getWeekPathPattern(week),
+            this.ctrl.config.getWeekFilePattern(week),
+        ]);
+        const path = J.Util.resolvePath(pathname.value!, filename.value!);
 
-            Promise.all([
-                this.ctrl.config.getWeekPathPattern(week),
-                this.ctrl.config.getWeekFilePattern(week)
-
-            ]).then(([pathname, filename]) => {
-                path = J.Util.resolvePath(pathname.value!, filename.value!);
-                return this.ctrl.ui.openDocument(path);
-
-            }).catch((reason: any) => {
-                if (reason instanceof Error) {
-                    if (!reason.message.startsWith("cannot open file:")) {
-                        this.ctrl.logger.printError(reason);
-                        reject(reason);
-                    }
-                }
-                return this.ctrl.writer.createWeeklyForPath(path, week);
-
-            }).then((_doc: vscode.TextDocument) => {
-                this.ctrl.logger.debug("loadEntryForWeek() - Loaded file in:", _doc.uri.toString());
-                resolve(_doc);
-
-            }).catch((error: Error) => {
-                this.ctrl.logger.printError(error);
-                reject("Failed to load entry for week: " + week);
-            });
-        });
+        const doc = await this.openOrCreate(
+            path,
+            () => this.ctrl.writer.createWeeklyForPath(path, week),
+        );
+        this.ctrl.logger.debug("loadEntryForWeek() - Loaded file in:", doc.uri.toString());
+        return doc;
     }
 
 
     /**
-     * Loads the journal entry for the given date. If no entry exists, promise is rejected with the invalid path
+     * Loads the journal entry for the given date. If no entry exists, it is created.
      *
      * @param {Date} date the date for the entry
-     * @returns {Q.Promise<vscode.TextDocument>} the document
-     * @throws {string} error message
+     * @returns {Promise<vscode.TextDocument>} the document
      * @memberof Reader
      */
-     public async loadEntryForDay(date: Date): Promise<vscode.TextDocument> {
+    public async loadEntryForDay(date: Date): Promise<vscode.TextDocument> {
+        if (J.Util.isNullOrUndefined(date) || date!.toString().includes("Invalid")) {
+            throw new Error("Invalid date");
+        }
+        this.ctrl.logger.trace("Entering loadEntryforDate() in actions/reader.ts for date " + date.toISOString());
 
-        return new Promise<vscode.TextDocument>((resolve, reject) => {
-            if (J.Util.isNullOrUndefined(date) || date!.toString().includes("Invalid")) {
-                reject("Invalid date");
-                return;
-            }
+        const [pathname, filename] = await Promise.all([
+            this.ctrl.config.getResolvedEntryPath(date),
+            this.ctrl.config.getEntryFilePattern(date),
+        ]);
+        const path = J.Util.resolvePath(pathname.value!, filename.value!);
 
-            this.ctrl.logger.trace("Entering loadEntryforDate() in actions/reader.ts for date " + date.toISOString());
+        const doc = await this.openOrCreate(
+            path,
+            () => this.ctrl.writer.createEntryForPath(path, date),
+        );
+        this.ctrl.logger.debug("loadEntryForDate() - Loaded file in:", doc.uri.toString());
 
-            let path: string = "";
+        new J.Provider.SyncNoteLinks(this.ctrl).injectAttachementLinks(doc, date)
+            .finally(() => this.ctrl.logger.trace("Scanning notes completed"));
 
-            Promise.all([
-                this.ctrl.config.getResolvedEntryPath(date),
-                this.ctrl.config.getEntryFilePattern(date)
-
-            ]).then(([pathname, filename]) => {
-                path = J.Util.resolvePath(pathname.value!, filename.value!);
-                return this.ctrl.ui.openDocument(path);
-
-
-            }).catch((reason: any) => {
-                if (reason instanceof Error) {
-                    if (!reason.message.startsWith("cannot open file:") && !reason.message.startsWith("cannot open vscode-remote:")) {
-                        this.ctrl.logger.printError(reason);
-                        reject(reason);
-                    }
-                }
-                return this.ctrl.writer.createEntryForPath(path, date);
-
-            }).then((_doc: vscode.TextDocument) => {
-                this.ctrl.logger.debug("loadEntryForDate() - Loaded file in:", _doc.uri.toString());
-                new J.Provider.SyncNoteLinks(this.ctrl).injectAttachementLinks(_doc, date)
-                    .finally(() =>
-                        // do nothing
-                        this.ctrl.logger.trace("Scanning notes completed")
-                    );
-                resolve(_doc);
-
-            }).catch((error: Error) => {
-                this.ctrl.logger.printError(error);
-                reject("Failed to load entry for date: " + date.toDateString());
-
-            });
-
-        });
+        return doc;
     }
 
+    private async openOrCreate(
+        path: string,
+        create: () => Promise<vscode.TextDocument>,
+    ): Promise<vscode.TextDocument> {
+        const exists = await J.Util.fileExists(vscode.Uri.file(path));
+        if (exists) {
+            return this.ctrl.ui.openDocument(path);
+        }
+        return create();
+    }
 }
 
 

--- a/src/provider/features/load-note.ts
+++ b/src/provider/features/load-note.ts
@@ -32,31 +32,21 @@ export class LoadNotes {
     } 
     
       /**
-     * Creates or loads a note 
+     * Creates or loads a note
      *
      * @param {string} path
      * @param {string} content
      * @returns {Promise<vscode.TextDocument>}
      * @memberof Writer
      */
-       public async loadNote(path: string, content: string): Promise<vscode.TextDocument> {
+    public async loadNote(path: string, content: string): Promise<vscode.TextDocument> {
         this.ctrl.logger.trace("Entering loadNote() in  features/load-note.ts for path: ", path);
 
-        return new Promise<vscode.TextDocument>((resolve, reject) => {
-            // check if file exists already
-
-            this.ctrl.ui.openDocument(path)
-                .then((doc: vscode.TextDocument) => resolve(doc))
-                .catch(error => {
-                    this.ctrl.writer.createSaveLoadTextDocument(path, content)
-                        .then((doc: vscode.TextDocument) => resolve(doc))
-                        .catch(error => {
-                            this.ctrl.logger.error(error);
-                            reject("Failed to load note.");
-                        });
-                });
-
-        });
+        const exists = await J.Util.fileExists(vscode.Uri.file(path));
+        if (exists) {
+            return this.ctrl.ui.openDocument(path);
+        }
+        return this.ctrl.writer.createSaveLoadTextDocument(path, content);
     }
     
 

--- a/src/test/suite/issue-51-remote-create.test.ts
+++ b/src/test/suite/issue-51-remote-create.test.ts
@@ -1,0 +1,111 @@
+import * as assert from 'assert';
+import * as os from 'os';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as J from '../..';
+import { LoadNotes } from '../../provider';
+import { fileExists } from '../../util/fs-exists';
+import { TestLogger } from '../test-logger';
+
+suite('Issue #51 — Stat-first file creation on remote workspaces', () => {
+
+    suite('fileExists helper', () => {
+
+        test('returns false for a missing path (no error logged)', async () => {
+            const missing = vscode.Uri.file(path.join(os.tmpdir(), `issue51-missing-${Date.now()}`));
+            const result = await fileExists(missing);
+            assert.strictEqual(result, false);
+        });
+
+        test('returns true for an existing path', async () => {
+            const target = vscode.Uri.file(path.join(os.tmpdir(), `issue51-exists-${Date.now()}.md`));
+            await vscode.workspace.fs.writeFile(target, new TextEncoder().encode('content'));
+            try {
+                const result = await fileExists(target);
+                assert.strictEqual(result, true);
+            } finally {
+                try { await vscode.workspace.fs.delete(target); } catch { /* ignore */ }
+            }
+        });
+    });
+
+    suite('open-before-create antipattern is gone', () => {
+        let originalBase: string | undefined;
+        let tmpBase: string;
+        let ctrl: J.Util.Ctrl;
+        let logger: TestLogger;
+        let openCallCount: number;
+
+        setup(async () => {
+            const config = vscode.workspace.getConfiguration('journal');
+            originalBase = config.get<string>('base');
+
+            tmpBase = path.join(os.tmpdir(), `issue51-base-${Date.now()}`);
+            await vscode.workspace.fs.createDirectory(vscode.Uri.file(tmpBase));
+            await config.update('base', tmpBase, vscode.ConfigurationTarget.Workspace);
+
+            const refreshedConfig = vscode.workspace.getConfiguration('journal');
+            ctrl = new J.Util.Ctrl(refreshedConfig);
+            logger = new TestLogger(false);
+            ctrl.logger = logger;
+
+            openCallCount = 0;
+            const original = ctrl.ui.openDocument.bind(ctrl.ui);
+            (ctrl.ui as any).openDocument = async (p: string | vscode.Uri) => {
+                openCallCount++;
+                return original(p);
+            };
+        });
+
+        teardown(async () => {
+            const config = vscode.workspace.getConfiguration('journal');
+            await config.update('base', originalBase, vscode.ConfigurationTarget.Workspace);
+            try {
+                await vscode.workspace.fs.delete(vscode.Uri.file(tmpBase), { recursive: true });
+            } catch { /* ignore */ }
+        });
+
+        test('loadEntryForDay does not call openDocument when the entry is missing', async () => {
+            const date = new Date();
+            const doc = await ctrl.reader.loadEntryForDay(date);
+            assert.ok(doc, 'expected a document');
+            assert.strictEqual(openCallCount, 0, 'openDocument must not be called on the create path');
+            assert.deepStrictEqual(logger.errors, [], 'no errors must be logged on the create path');
+        });
+
+        test('loadEntryForDay opens an existing entry without overwriting its content', async () => {
+            const date = new Date();
+            const first = await ctrl.reader.loadEntryForDay(date);
+            const entryPath = first.uri.fsPath;
+
+            const sentinel = `# sentinel ${Date.now()}\n`;
+            await vscode.workspace.fs.writeFile(vscode.Uri.file(entryPath), new TextEncoder().encode(sentinel));
+
+            openCallCount = 0;
+            logger.errors.length = 0;
+            await ctrl.reader.loadEntryForDay(date);
+
+            const onDisk = new TextDecoder().decode(await vscode.workspace.fs.readFile(vscode.Uri.file(entryPath)));
+            assert.ok(onDisk.includes('sentinel'), 'existing content must be preserved');
+            assert.strictEqual(openCallCount, 1, 'openDocument must be called once for an existing file');
+            assert.deepStrictEqual(logger.errors, [], 'no errors must be logged when opening an existing file');
+        });
+
+        test('loadEntryForWeek does not call openDocument when the weekly entry is missing', async () => {
+            const doc = await ctrl.reader.loadEntryForWeek(20);
+            assert.ok(doc, 'expected a document');
+            assert.strictEqual(openCallCount, 0, 'openDocument must not be called on the create path');
+            assert.deepStrictEqual(logger.errors, [], 'no errors must be logged on the create path');
+        });
+
+        test('LoadNotes.loadNote does not call openDocument when the note is missing', async () => {
+            const notePath = path.join(tmpBase, `note-${Date.now()}.md`);
+            const input = new J.Model.NoteInput();
+            input.text = 'issue51 test note';
+            const doc = await new LoadNotes(input, ctrl).loadNote(notePath, '# Test\n');
+            assert.ok(doc, 'expected a document');
+            assert.strictEqual(openCallCount, 0, 'openDocument must not be called on the create path');
+            assert.deepStrictEqual(logger.errors, [], 'no errors must be logged on the create path');
+        });
+    });
+});

--- a/src/test/test-logger.ts
+++ b/src/test/test-logger.ts
@@ -1,25 +1,29 @@
 import { Logger } from "../util/logger";
 
 export class TestLogger implements Logger {
+    public readonly errors: string[] = [];
+
     constructor(public tracing: boolean) {
 
     }
 
     error(message: string, ...optionalParams: any[]): void {
-        console.error("ERROR", message, ...optionalParams); 
+        this.errors.push(String(message));
+        console.error("ERROR", message, ...optionalParams);
     }
     printError(error: Error): void {
-        throw new Error("Method not implemented.");
+        this.errors.push(error.message);
+        console.error("ERROR", error.message);
     }
     showChannel(): void {
-        throw new Error("Method not implemented.");
+        // no-op in tests
     }
     debug(message: string, ...optionalParams: any[]): void {
-        console.debug("DEBUG", message, ...optionalParams); 
+        console.debug("DEBUG", message, ...optionalParams);
     }
     trace(message: string, ...optionalParams: any[]): void {
         if(this.tracing) {
-            console.trace(message, ...optionalParams); 
+            console.trace(message, ...optionalParams);
         }
         // do nothing
 

--- a/src/util/fs-exists.ts
+++ b/src/util/fs-exists.ts
@@ -1,0 +1,35 @@
+// Copyright (C) 2026  Patrick Maué
+//
+// This file is part of vscode-journal.
+//
+// vscode-journal is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// vscode-journal is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with vscode-journal.  If not, see <http://www.gnu.org/licenses/>.
+
+'use strict';
+
+import * as vscode from 'vscode';
+
+// Returns true if the URI resolves to an existing entry on the workspace
+// filesystem, false if it definitely does not exist, and rethrows for any
+// other FS error so callers (and VS Code) can surface a notification.
+export async function fileExists(uri: vscode.Uri): Promise<boolean> {
+    try {
+        await vscode.workspace.fs.stat(uri);
+        return true;
+    } catch (err) {
+        if (err instanceof vscode.FileSystemError && err.code === 'FileNotFound') {
+            return false;
+        }
+        throw err;
+    }
+}

--- a/src/util/index.ts
+++ b/src/util/index.ts
@@ -52,3 +52,4 @@ export {
     inferType,
     resolvePath,
 } from './paths';
+export { fileExists } from './fs-exists';

